pkg/kubelet/volumemanager/reconciler/reconciler.go

```diff
@@ -521,17 +554,30 @@ func (rc *reconciler) updateStates(volumesNeedUpdate map[v1.UniqueVolumeName]*re
 			volume.mounter,
 			volume.outerVolumeSpecName,
 			volume.volumeGidValue)
+
 		if err != nil {
 			glog.Errorf("Could not add pod to volume information to actual state of world: %v", err)
 			continue
 		}
+
 		if volume.pluginIsAttachable {
-			err = rc.actualStateOfWorld.MarkDeviceAsMounted(volume.volumeName)
-			if err != nil {
-				glog.Errorf("Could not mark device is mounted to actual state of world: %v", err)
+			if isNotMount, err := rc.mounter.IsLikelyNotMountPoint(volume.deviceMountPath); err != nil {
+				glog.Errorf("Could not check whether the volume: %v device %v is mounted at path: %v, error: %v",
+					volume.volumeName,
+					volume.devicePath,
+					volume.deviceMountPath,
+					err)
 				continue
+			} else if !isNotMount {
+				err = rc.actualStateOfWorld.MarkDeviceAsMounted(volume.volumeName)
+				if err != nil {
+					glog.Errorf("Could not mark device is mounted to actual state of world: %v", err)
+					continue
+				}
+				glog.Infof("Volume: %v is mounted", volume.volumeName)
```

Comment:
I think we can put the check of mountpoint into reconstructVolume function. If it is not a mountpoint, fail to reconstruct volume directly. This way, no need to add deviceMountPath into the reconstructedVolume data structure

Reply:
Isn't it a odd place for doing a mount check? mount check has nothing to do with reconstruction right?

Reply:
the reason for reconstruction is mainly because when kubelet restarts, we don't have the cached state information. If there are mounts left on the node, we should reconstruct it. But if we checked that the mounts no longer exist, we can ignore it without reconstructing the volume back to actual state, I think.

Reply:
Ok. It is infact then "reconstructActualStateofVolume". It would work, the only minor concern is failing reconstruction because a volume is not mounted. Volume is an object and mount is one of the intermediate states of volume. We are combining both.

Reply:
yeah, the function name is not very clear. I am fine with changing it to reconstuctActualStateofVolume. We have to reconstruct the spec because it is needed in the actual state.